// telegraf/lib/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mention = exports.link = exports.pre = exports.code = exports.underline = exports.strikethrough = exports.spoiler = exports.italic = exports.bold = exports.fmt = exports.FmtString = void 0;
const formatting_1 = require("./core/helpers/formatting");
Object.defineProperty(exports, "FmtString", { enumerable: true, get: function () { return formatting_1.FmtString; } });
exports.fmt = (0, formatting_1._fmt)('very-plain');
exports.bold = (0, formatting_1._fmt)('bold');
exports.italic = (0, formatting_1._fmt)('italic');
exports.spoiler = (0, formatting_1._fmt)('spoiler');
exports.strikethrough = (0, formatting_1._fmt)('strikethrough');
exports.underline = (0, formatting_1._fmt)('underline');
exports.code = (0, formatting_1._fmt)('code');
const pre = (language) => (0, formatting_1._fmt)('pre', { language });
exports.pre = pre;
const link = (content, url) => (0, formatting_1.linkOrMention)(content, { type: 'text_link', url });
exports.link = link;
const mention = (name, user) => typeof user === 'number'
    ? (0, exports.link)(name, 'tg://user?id=' + user)
    : (0, formatting_1.linkOrMention)(name, { type: 'text_mention', user });
exports.mention = mention;
